cpp/src/arrow/builder.cc

```diff
@@ -1222,9 +1222,21 @@ Status BinaryBuilder::Init(int64_t elements) {
 Status BinaryBuilder::Resize(int64_t capacity) {
   DCHECK_LT(capacity, std::numeric_limits<int32_t>::max());
   // one more then requested for offsets
-  RETURN_NOT_OK(offsets_builder_.Resize((capacity + 1) * sizeof(int64_t)));
+  RETURN_NOT_OK(offsets_builder_.Resize((capacity + 1) * sizeof(int32_t)));
   return ArrayBuilder::Resize(capacity);
 }
+  
+Status BinaryBuilder::ReserveData(int64_t capacity) {
+  if (value_data_length() + capacity > data_capacity_) {
+    if (value_data_length() + capacity > std::numeric_limits<int32_t>::max()) {
+      return Status::Invalid("Cannot reserve capacity larger than 2^31 - 1 in length for binary data");
+    }    
+    
+    RETURN_NOT_OK(value_data_builder_.Resize(value_data_length() + capacity));
```

Comment:
If you rebase on master, you can use `value_data_builder_.Reserve(capacity)` here